### ISSUE TITLE
Caught exception in Test_deadly_signal_TERM(): Vim(call):E684: list index out of range: 1

### ISSUE THREAD

**Issue by llebout** [vim/vim#7402]
```
Executed:  3515 Tests
 Skipped:    75 Tests
  FAILED:     1 Tests


Failures: 
        From test_signals.vim:
        Found errors in Test_deadly_signal_TERM():
        Run 1:
        Caught exception in Test_deadly_signal_TERM(): Vim(call):E684: list index out of range: 1 @ command line..script /tmp/guix-build-vim-8.2.2067.drv-0/source/src/testdir/runtest.vim[468]..function RunTheTest[39]..Test_deadly_signal_TERM, line 40
        Run 2:
        Caught exception in Test_deadly_signal_TERM(): Vim(call):E684: list index out of range: 1 @ command line..script /tmp/guix-build-vim-8.2.2067.drv-0/source/src/testdir/runtest.vim[502]..function RunTheTest[39]..Test_deadly_signal_TERM, line 40
        Run 3:
        Caught exception in Test_deadly_signal_TERM(): Vim(call):E684: list index out of range: 1 @ command line..script /tmp/guix-build-vim-8.2.2067.drv-0/source/src/testdir/runtest.vim[502]..function RunTheTest[39]..Test_deadly_signal_TERM, line 40
        Flaky test failed too often, giving up

TEST FAILURE
```

GNU Guix version:
```
  guix 04b8367
    repository URL: https://git.savannah.gnu.org/git/guix.git
    branch: master
    commit: 04b83678653fda3c66e600e88f54f5108290ec1c
```

GNU Guix System Distribution version:
```
  file name: /var/guix/profiles/system-9-link
  canonical file name: /gnu/store/ddz1csfjw8prf6rjh8rw36b4rfaqxjw1-system
  label: GNU with Linux-Libre 5.9.8
  bootloader: grub
  root device: UUID: 56b95577-afbd-4769-88b5-7ad6dd5a7a7f
  kernel: /gnu/store/jrgkcfd5lxrljirwqkv29jcdipha2vkc-linux-libre-5.9.8/bzImage
  channels:
    guix:
      repository URL: https://git.savannah.gnu.org/git/guix.git
      branch: master
      commit: abb64b5efc6b657d64e903f3e1bf28f9e3031f6c
  configuration file: /gnu/store/f11647640i49wbk89wkz206avglamsgy-configuration.scm
```

/gnu/store/f11647640i49wbk89wkz206avglamsgy-configuration.scm
```scheme
;; This is an operating system configuration generated
;; by the graphical installer.

(use-modules (gnu))
(use-service-modules desktop networking ssh xorg spice)

(operating-system
  (locale "en_US.utf8")
  (timezone "Europe/Paris")
  (keyboard-layout (keyboard-layout "gb"))
  (host-name "guix-vm")
  (users (cons* (user-account
                  (name "user")
                  (comment "User")
                  (group "users")
                  (home-directory "/home/user")
                  (supplementary-groups
                    '("wheel" "netdev" "audio" "video")))
                %base-user-accounts))
  (packages
    (append
      (list (specification->package "nss-certs")
            (specification->package "glibc-utf8-locales"))
      %base-packages))
  (services
    (append
      (list (spice-vdagent-service)
            (service gnome-desktop-service-type)
            (set-xorg-configuration
              (xorg-configuration
                (keyboard-layout keyboard-layout))))
      %desktop-services))
  (bootloader
    (bootloader-configuration
      (bootloader grub-bootloader)
      (target "/dev/sda")
      (keyboard-layout keyboard-layout)))
  (swap-devices (list "/dev/sda2"))
  (file-systems
    (cons* (file-system
             (mount-point "/")
             (device
               (uuid "56b95577-afbd-4769-88b5-7ad6dd5a7a7f"
                     'ext4))
             (type "ext4"))
           %base-file-systems)))
```

Terminal:
GNOME Terminal 3.34.2
https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/gnome.scm?id=abb64b5efc6b657d64e903f3e1bf28f9e3031f6c#n5015

Package definition:
```scheme
(define-public gnome-terminal
  (package
    (name "gnome-terminal")
    (version "3.34.2")
    (source
     (origin
       (method url-fetch)
       (uri (string-append "mirror://gnome/sources/" name "/"
                           (version-major+minor version) "/"
                           name "-" version ".tar.xz"))
       (sha256
        (base32
         "0gc004f9b5k94gkdanmqjz3wqgnpny0l3nqm8zd19h4f0ps27mrv"))))
    (build-system glib-or-gtk-build-system)
    (arguments
     '(#:configure-flags
       (list "--disable-migration" "--disable-search-provider"
             "--without-nautilus-extension")
       #:phases
       (modify-phases %standard-phases
         (add-before 'configure 'patch-/bin/true
                     (lambda _
                       (substitute* "configure"
                         (("/bin/true") (which "true"))))))))
    (native-inputs
     `(("pkg-config" ,pkg-config)
       ("desktop-file-utils" ,desktop-file-utils)
       ("intltool" ,intltool)
       ("itstool" ,itstool)
       ("xmllint" ,libxml2)))
    (propagated-inputs
     `(("dconf" ,dconf)))
    (inputs
     `(("gtk+" ,gtk+)
       ("vte" ,vte)
       ("gnutls" ,gnutls)
       ("gsettings-desktop-schemas" ,gsettings-desktop-schemas)
       ("util-linux" ,util-linux "lib")
       ("vala" ,vala)))
    (home-page "https://wiki.gnome.org/Apps/Terminal")
    (synopsis "Terminal emulator")
    (description
     "GNOME Terminal is a terminal emulator application for accessing a
UNIX shell environment which can be used to run programs available on
your system.

It supports several profiles, multiple tabs and implements several
keyboard shortcuts.")
    (license license:gpl3+)))
```

vim package definition:
https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/vim.scm?id=04b83678653fda3c66e600e88f54f5108290ec1c#n70
```scheme
(define-public vim
  (package
    (name "vim")
    (version "8.2.2067")
    (source (origin
             (method git-fetch)
             (uri (git-reference
                    (url "https://github.com/vim/vim")
                    (commit (string-append "v" version))))
             (file-name (git-file-name name version))
             (sha256
              (base32
               "02cd953h69k9klrcwi756namwg39ka7if9ccc399ihb1l5f3kr66"))))
    (build-system gnu-build-system)
    (arguments
     `(#:test-target "test"
       #:parallel-tests? #f
       #:phases
       (modify-phases %standard-phases
         (add-after 'configure 'patch-absolute-paths
           (lambda _
             (substitute* "runtime/tools/mve.awk"
               (("/usr/bin/nawk") (which "gawk")))
             (substitute* '("src/testdir/Makefile"
                            "src/testdir/test_normal.vim"
                            "src/testdir/test_popupwin.vim"
                            "src/testdir/test_system.vim"
                            "src/testdir/test_terminal.vim"
                            "src/testdir/test_terminal2.vim")
               (("/bin/sh") (which "sh")))
             (substitute* "src/testdir/test_autocmd.vim"
               (("/bin/kill") (which "kill")))
             #t))
         (add-before 'check 'set-environment-variables
           (lambda* (#:key inputs #:allow-other-keys)
             ;; One of the tests tests timezone-dependent functions.
             (setenv "TZDIR"
                     (string-append (assoc-ref inputs "tzdata")
                                    "/share/zoneinfo"))

             ;; Make sure the TERM environment variable is set for the tests
             (setenv "TERM" "xterm")
             #t))
         (add-before 'check 'skip-or-fix-failing-tests
           (lambda _
             ;; This test assumes that PID 1 is run as root and that the user
             ;; running the test suite does not have permission to kill(1, 0)
             ;; it.  This is not true in the build container, where both PID 1
             ;; and the test suite are run as the same user.  Skip the test.
             ;; An alternative fix would be to patch the PID used to a random
             ;; 32-bit value and hope it never shows up in the test environment.
             (substitute* "src/testdir/test_swap.vim"
               (("if !IsRoot\\(\\)") "if 0"))

             ;; These tests compares output against a golden ‘…/|b|i|n|/|s|h…’
             ;; literal.  We need to match that and substitute a similarly
             ;; ‘spliced’ path to ‘sh’ in the store, truncated to the last
             ;; 44 (spliced: 88) or so characters.
             ;; Two of the tests we simply skip instead of patching the screen dump.
             (substitute* "src/testdir/test_popupwin.vim"
               ((".*Test_popupwin_term_0[1|2].*") ""))
             ;; We replace the external program call (!) with a scroll-back (<)
             ;; symbol and blindly fix some other differences based on error output.
             (let ((splice (lambda (s separator)
                               (string-join (map string (string->list s))
                                            separator))))
               (substitute* "src/testdir/dumps/Test_terminal_from_cmd.dump"
                 (((splice "/bin/sh" "\\|"))
                  (splice (string-take-right (which "sh") 44) "|"))
                 (("^\\|!") "|<")
                 (("@37") ""))
               (substitute* '("src/testdir/dumps/Test_terminal_scrollback_1.dump"
                              "src/testdir/dumps/Test_terminal_scrollback_2.dump")
                 (((splice "/bin/sh" "\\|"))
                  (splice (string-take-right (which "sh") 61) "|"))
                 (("^\\|!") "|<")
                 ((" @55") " @1"))
               (substitute* '("src/testdir/dumps/Test_terminal_scrollback_3.dump"
                              "src/testdir/dumps/Test_popupwin_term_03.dump"
                              "src/testdir/dumps/Test_popupwin_term_04.dump")
                 (((splice "/bin/sh" "\\|"))
                  (splice (string-take-right (which "sh") 62) "|"))
                 (("^\\|!") "|<")
                 (("\\]\\| @56") "]| @1")))
             #t)))))
    (inputs
     `(("gawk" ,gawk)
       ("ncurses" ,ncurses)
       ("perl" ,perl)
       ("tcsh" ,tcsh)))                 ; For runtime/tools/vim32
    (native-inputs
     `(("libtool" ,libtool)

       ;; For tests.
       ("tzdata" ,tzdata-for-tests)))
    (home-page "https://www.vim.org/")
    (synopsis "Text editor based on vi")
    ;; The description shares language with the vim-full package. When making
    ;; changes, check if the other description also needs to be updated.
    (description
     "Vim is a highly configurable text editor built to enable efficient text
editing.  It is an improved version of the vi editor distributed with most UNIX
systems.

Vim is often called a \"programmer's editor,\" and so useful for programming
that many consider it an entire IDE.  It's not just for programmers, though.
Vim is perfect for all kinds of text editing, from composing email to editing
configuration files.")
    (license license:vim)))
```

You can reproduce with:
* Install GNU Guix
* `$ guix pull --commit 04b83678653fda3c66e600e88f54f5108290ec1c`
* `$ guix build vim --keep-failed 2>vim-build.log`
* Use less to read the file otherwise termcaps mess the terminal

It also happens with the current latest version:
* `$ guix build --with-commit=vim=v8.2.2077 vim --keep-failed 2>vim-build.log`

**Comment by chrisbra** [comment]
is this a recent regression? It looks like the `VimLeave` autocommand is not triggered. Any thing special on GNU Guix related to signal processing? Oh can you also share the content of the `XautoOut` file?

**Comment by llebout** [comment]
@chrisbra Recent yes, what I know is that it didnt happen in 8.2.1980. About signals, I don't think so. It's just Linux?

```
user@guix-vm /tmp/guix-build-vim-8.2.2077.drv-0/source$ cat src/testdir/XautoOut 
VimLeavePre triggered
```

**Comment by chrisbra** [comment]
can you do a bisect between 8.2.1980 and current master?

**Comment by brammool** [comment]
Patch 8.2.2070 added setting v:exiting.  That's the closest I can think of related to VimLeave.
It should not have any effect like this, perhaps it's timing related?

**Comment by llebout** [comment]
I figured even v8.2.1980 has the bug and it also reproduces on Fedora 33 fully updated as of today.
I'm trying to bisect, taking a while.

**Comment by llebout** [comment]
@chrisbra @brammool how would you easily and reliably run a single test through the Makefile?

**Comment by dpelle** [comment]
Following patch won't fix anything but at least the information will be clearer when `Test_deadly_signal_TERM()` fails:
```
diff --git a/src/testdir/test_signals.vim b/src/testdir/test_signals.vim
index 78b2cecb8..1723f3310 100644
--- a/src/testdir/test_signals.vim
+++ b/src/testdir/test_signals.vim
@@ -153,8 +153,7 @@ func Test_deadly_signal_TERM()
   call assert_equal(['foo'], getline(1, '$'))
 
   let result = readfile('XautoOut')
-  call assert_match('VimLeavePre triggered', result[0])
-  call assert_match('VimLeave triggered', result[1])
+  call assert_equal(["VimLeavePre triggered", "VimLeave triggered"], result)
 
   %bwipe!
   call delete('.Xsig_TERM.swp')
```

> how would you easily and reliably run a single test through the Makefile

```
$ cd vim/src/testdir
$ make clean
$ make test_signals
```

**Comment by brammool** [comment]
If you say it didn't happen in 8.2.1980, and now that you bisect it does happen, that would point to a change in the compiler and/or libraries.

**Comment by llebout** [comment]
Thanks, I have first bad commit as:

```
commit 129d6bf6b3d120b0a4c69e18b5e8602a84e352bf (tag: v8.2.0769, refs/bisect/bad)
Author: Bram Moolenaar <Bram@vim.org>
Date:   Sat May 16 16:08:35 2020 +0200

    patch 8.2.0769: VimLeavePre not triggered when Vim is terminated
    
    Problem:    VimLeavePre not triggered when Vim is terminated.
    Solution:   Unblock autocommands.
```

**Comment by llebout** [comment]
> If you say it didn't happen in 8.2.1980, and now that you bisect it does happen, that would point to a change in the compiler and/or libraries.

I did not run tests on v8.2.1980 myself, GNU Guix had a substitute (pre-compiled) for v8.2.1980 which I had installed previously. Substitutes are not available unless tests pass though so I see two possible influences: GNU Guix's grafts or system configuration (e.g. kernel). Not sure how the current official GNU Guix substitute machine looks like. It should have never produced a substitute if tests didnt pass.

However, I have bisected on Fedora 33 now so let's forget GNU Guix.

**Comment by llebout** [comment]
@dpelle 

Test result with your patch (applied on first bad commit v8.2.0769):

```
$ make test_signals
rm -f test_signals.res test.log messages
rm -rf test1.failed test.out X* viminfo tiny.vim small.vim mbyte.vim mzscheme.vim test.ok benchmark.out wrongtermsize
VIMRUNTIME=../../runtime  ../vim -f  -u unix.vim -U NONE --noplugin --not-a-term -s dotest.in test1.in --cmd 'au SwapExists * let v:swapchoice = "e"' > /dev/null
rm -rf X* viminfo
VIMRUNTIME=../../runtime  ../vim -f  -u unix.vim -U NONE --noplugin --not-a-term -S runtest.vim test_signals.vim --cmd 'au SwapExists * let v:swapchoice = "e"' > /dev/null


From test_signals.vim:
Executed Test_deadly_signal_TERM() in   0.100508 seconds
Found errors in Test_deadly_signal_TERM():
function RunTheTest[39]..Test_deadly_signal_TERM line 35: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Flaky test failed, running it again
Executed Test_deadly_signal_TERM() in   0.145031 seconds
Found errors in Test_deadly_signal_TERM():
function RunTheTest[39]..Test_deadly_signal_TERM line 35: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Executed Test_signal_INT() in   0.140495 seconds
Executed Test_signal_PWR() in   0.013618 seconds
Executed Test_signal_WINCH() in   0.010123 seconds
Executed 5 tests in   2.411905 seconds
1 FAILED:
Found errors in Test_deadly_signal_TERM():
Run 1:
function RunTheTest[39]..Test_deadly_signal_TERM line 35: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Run 2:
function RunTheTest[39]..Test_deadly_signal_TERM line 35: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Flaky test failed too often, giving up
make: *** [Makefile:78: test_signals] Error 1
```

**Comment by llebout** [comment]
@dpelle Applied again current master:

```
$ make test_signals
rm -f test_signals.res test.log messages
VIMRUNTIME=../../runtime  ../vim -f  -u unix.vim -U NONE --noplugin --not-a-term -S runtest.vim test_signals.vim --cmd 'au SwapExists * let v:swapchoice = "e"' > /dev/null


From test_signals.vim:
Executed Test_deadly_signal_TERM()                 in   0.101221 seconds
Found errors in Test_deadly_signal_TERM():
function RunTheTest[39]..Test_deadly_signal_TERM line 39: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Flaky test failed, running it again
Executed Test_deadly_signal_TERM()                 in   0.124172 seconds
Found errors in Test_deadly_signal_TERM():
function RunTheTest[39]..Test_deadly_signal_TERM line 39: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Executed Test_signal_INT()                         in   0.139752 seconds
Executed Test_signal_PWR()                         in   0.007113 seconds
Executed Test_signal_WINCH()                       in   0.010039 seconds
Executed 5 tests                         in   2.384303 seconds
1 FAILED:
Found errors in Test_deadly_signal_TERM():
Run 1:
function RunTheTest[39]..Test_deadly_signal_TERM line 39: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Run 2:
function RunTheTest[39]..Test_deadly_signal_TERM line 39: Expected ['VimLeavePre triggered', 'VimLeave triggered'] but got []
Flaky test failed too often, giving up
make: *** [Makefile:66: test_signals] Error 1
```

**Comment by brammool** [comment]
Patch 8.2.0769 added the test, so obviously it didn't fail before that. This patch is from May, I doubt you wouldn't have seen the failure for half a year.  Again, I suspect something in the environment changed.

To run a single test: "make test_signals".  You can even narrow it down to one function by setting $TEST_FILTER to the name of the function.

**Comment by chrisbra** [comment]
Fedora 33 is quite new. Are there any obvious changes to kernel or build toolchain in it? something that is the same as in GUIX? How did you compile it? Can you manually reproduce the issue from this particular test?

**Comment by dpelle** [comment]
I now recall that this test `Test_deadly_signal_TERM` was flaky.
Although the failure was hard to reproduce on my Linux machine, the following patch was made to attempt to fix it, but perhaps it does not fix it completely?  I have not seen `Test_deadly_signal_TERM` failing in a long time though, at least not since 8.2.0874:

```
$ git show d14fd5285e491a39028c4b4722ddbe7c9dfa9bb2
commit d14fd5285e491a39028c4b4722ddbe7c9dfa9bb2 (tag: v8.2.0874)
Author: Bram Moolenaar <Bram@vim.org>
Date:   Mon Jun 1 15:05:19 2020 +0200

    patch 8.2.0874: signals test is a bit flaky
    
    Problem:    Signals test is a bit flaky.
    Solution:   Flush the XautoOut file.  Delete files that may be left behind
                from a failure. (Dominique Pelle, closes #6179)
```

**Comment by chrisbra** [comment]
> Flush the XautoOut file

This sounds like it might depend on the filesystem and its mount options.

**Comment by llebout** [comment]
@chrisbra I compile and test by cloning the repo and using: `$ ./configure && make -j$(nproc) && make test` - simple as that. GNU Guix has Glibc 2.31 currently, Fedora Glibc 2.32. Otherwise Fedora 33 has Linux 5.9.10 currently, GNU Guix has Linux-libre 5.9.11 (but I also tried Linux-libre 5.4.x, same issue). Linux-libre is identical w.r.t. signal handling, it only strips proprietary blobs and firmware. GNU Guix uses GCC 7.x currently and Fedora uses GCC 10.x. I have ext4 file systems on both GNU Guix and Fedora systems. I'm still trying to find out what this is caused by.

**Comment by chrisbra** [comment]
what are your mount options for ext4?

**Comment by llebout** [comment]
@chrisbra Nevermind, I was compiling on tmpfs so not relevant. `tmpfs on /tmp type tmpfs (rw,nosuid,nodev,seclabel,nr_inodes=409600,inode64)`. I also just tried compiling in my home directory that's ext4 and same result. `/dev/mapper/fedora_localhost--live-home on /home type ext4 (rw,relatime,seclabel)`

If you try with Fedora 33 in a VM or container, does that bug really not happen to you?

**Comment by chrisbra** [comment]
I don't have a Fedora 33 available yet, so I cannot test. In any case I suspect this is a configuration issue on your system (or some optimization). If you mount using `data=journal` does that work? This may reduce performance by disableing some performance optimizations. so not sure. Perhaps also tweeking the 'updatetime' setting for that test is required so that the swap file is actually written by vim. Something like the following patch:

```patch
diff --git a/src/testdir/test_signals.vim b/src/testdir/test_signals.vim
index 1723f3310..db2bcd169 100644
--- a/src/testdir/test_signals.vim
+++ b/src/testdir/test_signals.vim
@@ -130,6 +130,7 @@ func Test_deadly_signal_TERM()
   call delete('XsetupAucmd')
   call delete('XautoOut')
   let lines =<< trim END
+    set updatetime=10
     au VimLeave * call writefile(["VimLeave triggered"], "XautoOut", "as")
     au VimLeavePre * call writefile(["VimLeavePre triggered"], "XautoOut", "as")
   END
```

**Comment by llebout** [comment]
@chrisbra How do you explain that it happens on tmpfs, how would it be caused by a configuration issue if it happens on both GNU Guix and Fedora? I am not sure I understand what kind of configuration issue you are talking about. Will be able to try the patch in few hours, thanks. 

For testing with Fedora 33 quickly you can use a container image. Say you got podman installed: `$ podman run -it fedora:33 /bin/bash`
This does not include some of the things a full Fedora 33 installation has but at least can reproduce software issues.

**Comment by dpelle** [comment]
When built with asan, `Test_deadly_signal_TERM()` succeeds most of the time but fails once in a while:
```
Executed:  3674 Tests
 Skipped:    49 Tests
  FAILED:     1 Tests


Failures: 
	From test_signals.vim:
	Found errors in Test_deadly_signal_TERM():
	Run 1:
	Caught exception in Test_deadly_signal_TERM(): Vim(let):E484: Can't open file XautoOut @ command line..script /home/pel/sb/vim/src/testdir/runtest.vim[468]..function RunTheTest[39]..Test_deadly_signal_TERM, line 38
	Run 2:
	Caught exception in Test_deadly_signal_TERM(): Vim(let):E484: Can't open file XautoOut @ command line..script /home/pel/sb/vim/src/testdir/runtest.vim[502]..function RunTheTest[39]..Test_deadly_signal_TERM, line 38
	Run 3:
	Caught exception in Test_deadly_signal_TERM(): Vim(let):E484: Can't open file XautoOut @ command line..script /home/pel/sb/vim/src/testdir/runtest.vim[502]..function RunTheTest[39]..Test_deadly_signal_TERM, line 38
	Flaky test failed too often, giving up

TEST FAILURE
Makefile:43: recipe for target 'report' failed
make: *** [report] Error 1
```
This happened with 8.2.2173 (huge, Linux x86_64, asan build).
Perhaps timing is involved (asan build is of course slower).

**Comment by brammool** [comment]
Dominique wrote:

> When built with asan, `Test_deadly_signal_TERM()` succeeds most of the time but fails once in a while:
> ```
> Executed:  3674 Tests
>  Skipped:    49 Tests
>   FAILED:     1 Tests
> 
> 
> Failures: 
> 	From test_signals.vim:
> 	Found errors in Test_deadly_signal_TERM():
> 	Run 1:
> 	Caught exception in Test_deadly_signal_TERM(): Vim(let):E484: Can't open file XautoOut @ command line..script /home/pel/sb/vim/src/testdir/runtest.vim[468]..function RunTheTest[39]..Test_deadly_signal_TERM, line 38
> 	Run 2:
> 	Caught exception in Test_deadly_signal_TERM(): Vim(let):E484: Can't open file XautoOut @ command line..script /home/pel/sb/vim/src/testdir/runtest.vim[502]..function RunTheTest[39]..Test_deadly_signal_TERM, line 38
> 	Run 3:
> 	Caught exception in Test_deadly_signal_TERM(): Vim(let):E484: Can't open file XautoOut @ command line..script /home/pel/sb/vim/src/testdir/runtest.vim[502]..function RunTheTest[39]..Test_deadly_signal_TERM, line 38
> 	Flaky test failed too often, giving up
> 
> TEST FAILURE
> Makefile:43: recipe for target 'report' failed
> make: *** [report] Error 1
> ```
> This happened with 8.2.2173 (huge, Linux x86_64, asan build).
> Perhaps timing is involved (asan build is of course slower).

The test is skipped with valgrind, because it may catch the TERM signal
and the behavior is different.  The test does seem to pass for me when I
comment out those lines though.  The INT signal test does fail under
valgrind.

Perhap ASAN also does something with the TERM signal?  That doesn't
explain why it would only fail once in a while though.

There are several WaitForAssert() calls, I don't see where it could be
flaky.

-- 
While it's true that many normal people whould prefer not to _date_ an
engineer, most normal people harbor an intense desire to _mate_ with them,
thus producing engineerlike children who will have high-paying jobs long
before losing their virginity.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
> Perhap ASAN also does something with the TERM signal? That doesn't
> explain why it would only fail once in a while though.

To clarify, the test `Test_deadly_signal_TERM()` succeeds the vast majority
of the time with asan.  But in rare cases it fails as described in my above comment.
I reproduced it on 2 different Linux Ubuntu-18.04 x86_64 machines.
The failure is rare though.  Note sure how frequent.
Interestingly, when it fails, it then retries 2 more times which always fails.

**Comment by dpelle** [comment]
I found something interesting: `Test_deadly_signal_TERM()` always fails when
uncommenting out 2 lines in `src/Makefile`:
```
diff --git a/src/Makefile b/src/Makefile
index c30af8b29..cb07e9173 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -612,7 +612,7 @@ CClink = $(CC)
 # Note: If you use -Wextra and get warnings in GTK code about function
 #       parameters, you can add -Wno-cast-function-type (but not with clang)
 #CFLAGS = -g -Wall -Wextra -Wshadow -Wmissing-prototypes -Wunreachable-code -Wno-cast-function-type -Wno-deprecated-declarations -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
-#CFLAGS = -g -Wall -Wextra -Wshadow -Wmissing-prototypes -Wunreachable-code -Wno-deprecated-declarations -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
+CFLAGS = -g -Wall -Wextra -Wshadow -Wmissing-prototypes -Wunreachable-code -Wno-deprecated-declarations -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
 # Add -Wpedantic to find // comments and other C99 constructs.
 # Better disable Perl and Python to avoid a lot of warnings.
 #CFLAGS = -g -Wall -Wextra -Wshadow -Wmissing-prototypes -Wpedantic -Wunreachable-code -Wunused-result -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
@@ -698,7 +698,7 @@ LINT_OPTIONS = -beprxzF
 # $ export ASAN_OPTIONS="print_stacktrace=1 log_path=asan"
 # $ export LSAN_OPTIONS="suppressions=`pwd`/testdir/lsan-suppress.txt"
 # When running tests output can be found in testdir/asan.*
-#SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
+SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
 #SANITIZER_CFLAGS = -g -O0 -fsanitize=undefined -fno-omit-frame-pointer
 
 # Similarly when compiling with clang and using ubsan.
```

But the same test passes when uncommenting only 1 line:
```
diff --git a/src/Makefile b/src/Makefile
index c30af8b29..7959c370f 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -698,7 +698,7 @@ LINT_OPTIONS = -beprxzF
 # $ export ASAN_OPTIONS="print_stacktrace=1 log_path=asan"
 # $ export LSAN_OPTIONS="suppressions=`pwd`/testdir/lsan-suppress.txt"
 # When running tests output can be found in testdir/asan.*
-#SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
+SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
 #SANITIZER_CFLAGS = -g -O0 -fsanitize=undefined -fno-omit-frame-pointer
 
 # Similarly when compiling with clang and using ubsan.
```

What makes it fails is that when vim is built without `-D_REENTRANT`, then
`make test_signals` always fails. Whereas with `-D_REENTRANT`, it passes.
Well, maybe there is still a rare even with `-D_REENTRANT` as I recall seeing
it failing rarely, but I could not reproduce the failure with `-D_REENTRANT`
at least today.

To bug submitter @leo-lb : does adding `-D_REENTRANT` fixes your issue?

**Comment by chrisbra** [comment]
@leo-lb any news?

**Comment by llebout** [comment]
@chrisbra @dpelle 

Sorry for the absence.

> To bug submitter @leo-lb : does adding -D_REENTRANT fixes your issue?

Yes it did! Thanks a lot for your research! I tried 3 times by now and it still doesnt fail.

**Comment by brammool** [comment]
Reviving this -  does using -D_REENTRANT require a configure check, or could we just add it always?
